cmd/contour/shutdownmanager.go

```diff
@@ -0,0 +1,132 @@
+	prometheusURL := fmt.Sprintf("http://%s:%d%s", s.envoyHost, s.envoyPort, s.prometheusPath)
+	envoyAdminURL := fmt.Sprintf("http://%s:%d/healthcheck/fail", s.envoyHost, s.envoyPort)
+
+	// Send shutdown signal to Envoy to start draining connections
+	err := shutdownEnvoy(envoyAdminURL)
+	if err != nil {
+		s.Errorf("Error sending envoy healthcheck fail: %v", err)
+	}
+
+	s.Infof("Sent healthcheck fail to Envoy...waiting %s before polling for draining connections", s.checkDelay)
+	time.Sleep(s.checkDelay)
+
+	for {
+		openConnections, err := getOpenConnections(prometheusURL, s.prometheusStat, s.prometheusValues)
+		if err != nil {
+			s.Error(err)
+		} else {
+			if openConnections <= s.minOpenConnections {
+				s.Infof("Found %d open connections with min number of %d connections. Shutting down...", openConnections, s.minOpenConnections)
+				return
+			}
+			s.Infof("Found %d open connections with min number of %d connections. Polling again...", openConnections, s.minOpenConnections)
+		}
+		time.Sleep(s.checkInterval)
+	}
+}
+
+// shutdownEnvoy sends a POST request to /healthcheck/fail to tell Envoy to start draining connections
+func shutdownEnvoy(url string) error {
+	resp, err := http.Post(url, "", nil)
+	if err != nil {
+		return fmt.Errorf("creating POST request for URL %q failed: %s", url, err)
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("POST request for URL %q returned HTTP status %s", url, resp.Status)
+	}
+	return nil
+}
+
+// getOpenConnections parses a http request to a prometheus endpoint returning the sum of values found
+func getOpenConnections(url, prometheusStat string, prometheusValues []string) (int, error) {
+	// Make request to Envoy Prometheus endpoint
+	resp, err := http.Get(url)
+	if err != nil {
+		return -1, fmt.Errorf("GET request for URL %q failed: %s", url, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return -1, fmt.Errorf("GET request for URL %q returned HTTP status %s", url, resp.Status)
+	}
+
+	// Parse Prometheus listener stats for open connections
+	return metrics.ParseOpenConnections(resp.Body, prometheusStat, prometheusValues)
+}
+
+func doShutdownManager(config *shutdownmanagerContext) error {
+	var g workgroup.Group
+
+	g.Add(func(stop <-chan struct{}) error {
+		config.Info("started envoy shutdown manager")
+		defer config.Info("stopped")
+
+		http.HandleFunc("/healthz", config.healthzHandler)
+		http.HandleFunc("/shutdown", config.shutdownHandler)
+		log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", config.httpServePort), nil))
+
+		return nil
+	})
+
+	return g.Run()
+}
+
+// registerShutdownManager registers the envoy shutdown sub-command and flags
+func registerShutdownManager(cmd *kingpin.CmdClause, log logrus.FieldLogger) (*kingpin.CmdClause, *shutdownmanagerContext) {
+	ctx := &shutdownmanagerContext{
+		FieldLogger: log,
+	}
+	shutdownmgr := cmd.Command("shutdown-manager", "Start envoy shutdown-manager.")
```

Comment:
In `serve.go` and `servecontext.go`, we use a `newServeContext()` function, do you think we should do this the same way instead of using the Kingpin defaults? I feel like maybe the function is a little easier place to find the defaults.

Reply:
I have them match now per your suggestion. Good point on making them consistent.